### PR TITLE
Pyside6 merge

### LINKED ISSUE #2257: move to Qt6
Qt6 is slowly maturing, and it is probably good time to start looking at switching versions.
I don't expect it to be much more troublesome than the previous qt4->qt5 transition.
More details:
https://www.pythonguis.com/faq/pyqt5-vs-pyqt6/

Also, we might want to think about switching to PySide2.

### PR DESCRIPTION
**Pull request by rozyczko**
## Description

PyQt5 got replaced by Pyside6. This is needed for maintainability.

This is a big change from the point of view of codebase and requirement changes.
Please test and review carefully.

When testing the dev build please remember to **compile GUI elements** (.ui -> .py) by running e.g. `python setup.py build`

Fixes #2257

## How Has This Been Tested?

Local development version
Windows installer.

## Review Checklist (please remove items if they don't apply):

- [x] Code has been reviewed
- [x] Functionality has been tested
- [X] Windows installer (GH artifact) has been tested (installed and worked) 
- [ ] MacOSX installer (GH artifact) has been tested (installed and worked) 
- [ ] User documentation is available and complete (if required)
- [ ] Developers documentation is available and complete (if required)
- [ ] The introduced changes comply with SasView license (BSD 3-Clause)

### CODE REVIEW

File: src/sas/qtgui/MainWindow/GuiManager.py
Comment (lucas-wilkins) [inline src/sas/qtgui/MainWindow/GuiManager.py:427]: Hmmmm, it's clearly to do with FittingWindow inheriting from ABC ulitimately, I think it might be to do with the metaclass (or lack thereof) in Perspecive.

File: src/sas/qtgui/Utilities/PreferencesPanel.py
Comment (lucas-wilkins) [inline src/sas/qtgui/Utilities/PreferencesPanel.py:1]: Why does git think that this code is all entirely new? Think something might have gone wrong somewhere.

**Review by lucas-wilkins — APPROVED** [review]
Looks good, but I've not done a functionality test.

File: src/sas/qtgui/Utilities/PreferencesPanel.py
Comment (rozyczko) [reply src/sas/qtgui/Utilities/PreferencesPanel.py:1]: probably while merging `main` back and manually resolving conflict?

**Comment by lucas-wilkins** [comment]
Currently this is failing to run on my machine:

> Process finished with exit code -1073740791 (0xC0000409)

I'm not having much luck getting more details by setting the exception hook.

**Comment by lucas-wilkins** [comment]
The issue appears to be here:

> icon.addPixmap(QtGui.QPixmap(":/res/ball.ico"), QtGui.QIcon.Normal, QtGui.QIcon.Off)

Perhaps I need to regenerate all the Qt things with this branch.

**Comment by lucas-wilkins** [comment]
Yeah, that was it. Probably need to let the developers know that they'll need to do that too!

**Comment by lucas-wilkins** [comment]
Also, running convertUI I got the following error messages, at least some are not pyside related, but some might be:

> Generating .\Calculators\UI\DataOperationUtilityUI.py ...
Generating .\Calculators\UI\DensityPanel.py ...
Generating .\Calculators\UI\GenericScatteringCalculator.py ...
Error: [SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\GenericScatteringCalculator.ui: Warning: The name 'line' (Line) is already in use, defaulting to 'line1'.
[SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\GenericScatteringCalculator.ui: Warning: The name 'groupBox_7' (QGroupBox) is already in use, defaulting to 'groupBox_71'.
[SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\GenericScatteringCalculator.ui: Warning: The name 'gridLayout_14' (QGridLayout) is already in use, defaulting to 'gridLayout_141'.

while executing '[USER_DIR]\.conda\envs\sasview\Lib\site-packages\PySide6\uic -g python [SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\GenericScatteringCalculator.ui -o [SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\GenericScatteringCalculator.py'
Generating .\Calculators\UI\KiessigPanel.py ...
Generating .\Calculators\UI\ResolutionCalculatorPanelUI.py ...
Error: [SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\ResolutionCalculatorPanelUI.ui: Warning: Z-order assignment: 'verticalSpacer_2' is not a valid widget.

while executing '[USER_DIR]\.conda\envs\sasview\Lib\site-packages\PySide6\uic -g python [SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\ResolutionCalculatorPanelUI.ui -o [SASVIEW_PATH]\sasview\src\sas\qtgui\Calculators\UI\ResolutionCalculatorPanelUI.py'
Generating .\Calculators\UI\SldPanel.py ...
Generating .\Calculators\UI\SlitSizeCalculator.py ...
Generating .\MainWindow\UI\AboutUI.py ...
Generating .\MainWindow\UI\AcknowledgementsUI.py ...
Generating .\MainWindow\UI\CategoryManagerUI.py ...
Generating .\MainWindow\UI\ChangeCategoryUI.py ...
Error: [SASVIEW_PATH]\sasview\src\sas\qtgui\MainWindow\UI\ChangeCategoryUI.ui: Warning: Z-order assignment: 'horizontalSpacer_2' is not a valid widget.

while executing '[USER_DIR]\.conda\envs\sasview\Lib\site-packages\PySide6\uic -g python [SASVIEW_PATH]\sasview\src\sas\qtgui\MainWindow\UI\ChangeCategoryUI.ui -o [SASVIEW_PATH]\sasview\src\sas\qtgui\MainWindow\UI\ChangeCategoryUI.py'
Generating .\MainWindow\UI\ChangeNameUI.py ...
Generating .\MainWindow\UI\DataExplorerUI.py ...
Generating .\MainWindow\UI\MainWindowUI.py ...
Generating .\MainWindow\UI\TabbedFileLoadUI.py ...
Generating .\MainWindow\UI\WelcomePanelUI.py ...
Generating .\Perspectives\Corfunc\UI\CorfuncPanel.py ...
Generating .\Perspectives\Corfunc\UI\SaveExtrapolated.py ...
Generating .\Perspectives\Fitting\UI\ComplexConstraintUI.py ...
Generating .\Perspectives\Fitting\UI\ConstraintWidgetUI.py ...
Generating .\Perspectives\Fitting\UI\FittingOptionsUI.py ...
Generating .\Perspectives\Fitting\UI\FittingWidgetUI.py ...
Generating .\Perspectives\Fitting\UI\GPUOptionsUI.py ...
Generating .\Perspectives\Fitting\UI\GPUTestResultsUI.py ...
Generating .\Perspectives\Fitting\UI\MultiConstraintUI.py ...
Generating .\Perspectives\Fitting\UI\OptionsWidgetUI.py ...
Generating .\Perspectives\Fitting\UI\OrderWidgetUI.py ...
Generating .\Perspectives\Fitting\UI\SmearingWidgetUI.py ...
Generating .\Perspectives\Invariant\UI\InvariantDetailsUI.py ...
Generating .\Perspectives\Invariant\UI\TabbedInvariantUI.py ...
Generating .\Perspectives\Inversion\UI\DMaxExplorer.py ...
Generating .\Perspectives\Inversion\UI\TabbedInversionUI.py ...
Generating .\Plotting\UI\AddTextUI.py ...
Generating .\Plotting\UI\BoxSumUI.py ...
Generating .\Plotting\UI\ColorMapUI.py ...
Error: [SASVIEW_PATH]\sasview\src\sas\qtgui\Plotting\UI\ColorMapUI.ui: Warning: Z-order assignment: 'verticalSpacer' is not a valid widget.

while executing '[USER_DIR]\.conda\envs\sasview\Lib\site-packages\PySide6\uic -g python [SASVIEW_PATH]\sasview\src\sas\qtgui\Plotting\UI\ColorMapUI.ui -o [SASVIEW_PATH]\sasview\src\sas\qtgui\Plotting\UI\ColorMapUI.py'
Generating .\Plotting\UI\LinearFitUI.py ...
Error: [SASVIEW_PATH]\sasview\src\sas\qtgui\Plotting\UI\LinearFitUI.ui: Warning: Z-order assignment: 'verticalSpacer' is not a valid widget.

while executing '[USER_DIR]\.conda\envs\sasview\Lib\site-packages\PySide6\uic -g python [SASVIEW_PATH]\sasview\src\sas\qtgui\Plotting\UI\LinearFitUI.ui -o [SASVIEW_PATH]\sasview\src\sas\qtgui\Plotting\UI\LinearFitUI.py'
Generating .\Plotting\UI\MaskEditorUI.py ...
Generating .\Plotting\UI\PlotLabelPropertiesUI.py ...
Generating .\Plotting\UI\PlotPropertiesUI.py ...
Generating .\Plotting\UI\ScalePropertiesUI.py ...
Generating .\Plotting\UI\SetGraphRangeUI.py ...
Generating .\Plotting\UI\SlicerParametersUI.py ...
Generating .\Plotting\UI\WindowTitleUI.py ...
Generating .\Utilities\OrientationViewer\UI\OrientationViewerControllerUI.py ...
Generating .\Utilities\Preferences\UI\PreferencesUI.py ...
Generating .\Utilities\Reports\UI\ReportDialogUI.py ...
Generating .\Utilities\UI\AddMultEditorUI.py ...
Generating .\Utilities\UI\FileConverterUI.py ...
Generating .\Utilities\UI\FrameSelectUI.py ...
Generating .\Utilities\UI\GridPanelUI.py ...
Generating .\Utilities\UI\ImageViewerOptionsUI.py ...
Error: [SASVIEW_PATH]\sasview\src\sas\qtgui\Utilities\UI\ImageViewerOptionsUI.ui: Warning: Z-order assignment: 'horizontalSpacer' is not a valid widget.

while executing '[USER_DIR]\.conda\envs\sasview\Lib\site-packages\PySide6\uic -g python [SASVIEW_PATH]\sasview\src\sas\qtgui\Utilities\UI\ImageViewerOptionsUI.ui -o [SASVIEW_PATH]\sasview\src\sas\qtgui\Utilities\UI\ImageViewerOptionsUI.py'
Generating .\Utilities\UI\ImageViewerUI.py ...
Generating .\Utilities\UI\ModelEditor.py ...
Generating .\Utilities\UI\PluginDefinitionUI.py ...
Generating .\Utilities\UI\PluginManagerUI.py ...
Generating .\Utilities\UI\PreferencesUI.py ...
Generating .\Utilities\UI\TabbedModelEditor.py ...
Generating [SASVIEW_PATH]\sasview\src\sas\qtgui\UI\main_resources_rc.py ...
Generating [SASVIEW_PATH]\sasview\src\sas\qtgui\UI\images_rc.py ...

**Comment by lucas-wilkins** [comment]
Can I also recommend that this is merged immediately into every open PR (and perhaps any branches being worked on), for the reasons mentioned before, the convertUI tool doesn't seem to replace everything that it should, and going back and forth between different versions will be annoying.

**Comment by rozyczko** [comment]
Yes, I can see these Qt warnings but they don't seem to affect anything (at least not immediately

File: src/sas/qtgui/Plotting/PlotterBase.py
Comment (krzywon) [inline src/sas/qtgui/Plotting/PlotterBase.py:278]: No longer needed? Maybe get rid of it.

File: .github/workflows/ci.yml
Comment (krzywon) [inline .github/workflows/ci.yml:161]: Why are the GUI tests disabled?

**Review by krzywon — COMMENTED** [review]
This looks fine. I have a few comments, none of which are show-stoppers. I haven't tested the functionality yet.

File: build_tools/requirements.txt
Comment (rozyczko) [reply build_tools/requirements.txt:37]: Yes. More recent versions of PySide6 have an issue with pyinstaller built executables :/

File: src/sas/qtgui/MainWindow/PackageGatherer.py
Comment (rozyczko) [reply src/sas/qtgui/MainWindow/PackageGatherer.py:123]: One queries PySide6 version, the other queries Qt version. With PyQt they could be different but with PySide they are the same, so I think you're right. One query is enough

File: src/sas/qtgui/Plotting/PlotterBase.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/PlotterBase.py:278]: This was left for introspection but I should at least add a comment.

File: .github/workflows/ci.yml
Comment (rozyczko) [reply .github/workflows/ci.yml:161]: Because some tests were failing on GH. After merging, I'll look closer at fixing them

**Comment by lucas-wilkins** [comment]
> @lucas-wilkins does this mean you have now done the functionality testing and happy to merge? or should we do a bit more testing?

Yeah, I approved it a while back.

Devs need to be told about the need to rebuild the UI files, and someone should update all open PRs when its merged.

**Comment by krzywon** [comment]
I'm seeing the following error when running from both the Windows installer and `run.py`:

> Traceback (most recent call last):
  File "sasview.py", line 15, in <module>
  File "sas\cli.py", line 153, in main
  File "sas\qtgui\MainWindow\MainWindow.py", line 105, in run_sasview
  screen_resolution = screen.geometry().center()
AttributeError: 'NoneType' object has no attribute 'geometry'

File: src/sas/qtgui/MainWindow/MainWindow.py
Comment (krzywon) [inline src/sas/qtgui/MainWindow/MainWindow.py:110]: As a follow-up, removing these lines and passing `None` to the main sasview window resolves the error I am seeing. `screen_resolution` is no longer being used in the `MainSasViewWindow` and could probably be removed.

```suggestion
    mainwindow = MainSasViewWindow(None)
```

**Comment by rozyczko** [comment]
@krzywon - do you get similar error when trying to run 5.0.5 rc1?
`screenAt` can be None, according to the docs
```
Returns the screen at point, or nullptr if outside of any screen.

```
maybe you're trying to open SV on a screen which is no longer attached?

**Comment by krzywon** [comment]
Here is how I am able to reliably reproduce the error:

1. Use multiple monitors.
2. Run the installed sasview.exe.
3. As soon as the splash window opens, move your cursor to a screen that SasView is not in.
4. Error is thrown.

I was seeing the error because the window SasView opened in was different than the one I launched it from.

**Comment by krzywon** [comment]
Note: The application launches if my cursor is in the same window as the app.

**Comment by rozyczko** [comment]
I tried this with both installed version ( https://github.com/SasView/sasview/actions/runs/4578212203 ) and local develop.

Two monitor setup.

The app will reliably open on the monitor when my cursor is placed at the QApplication init time. I.e.
With the splash screen visible, moving the cursor to another monitor causes SasView to open on that monitor's screen.
Leaving the cursor will make SV open in the same screen as the splash screen.

I tested this with SV starting from both monitors.

Tomorrow I will try a 3-monitor setup.

(Win 10/64)

**Comment by rozyczko** [comment]
Due to intermittency of the issue, removing the offending lines is probably a good idea. The resolution passed to the MainWindow class is unused and the window is maximised anyway.

**Review by krzywon — APPROVED** [review]
The latest change seems to have solved the issue.

**Comment by butlerpd** [comment]
This can be merged once the conflict caused by the previous merge is resolved.

**Pull request merged**
